src/metabase/upload.clj

```diff
@@ -608,35 +601,60 @@
   ;; Assumes table-cols are unique when normalized
   (let [normalized-field-names (keys fields-by-normed-name)
         normalized-header      (map normalize-column-name header)
-        [extra missing _both] (data/diff (set normalized-header) (set normalized-field-names))]
+        [_extra missing _both] (data/diff (set normalized-header) (set normalized-field-names))]
     ;; check for duplicates
     (when (some #(< 1 %) (vals (frequencies normalized-header)))
       (throw (ex-info (tru "The CSV file contains duplicate column names.")
                       {:status-code 422})))
-    (when (or extra missing)
-      (let [error-message (extra-and-missing-error-markdown extra missing)]
+    (when missing
+      (let [error-message (missing-error-markdown missing)]
         (throw (ex-info error-message {:status-code 422}))))))
 
 (defn- matching-or-upgradable? [current-type relaxed-type]
-  (or (= current-type relaxed-type)
+  (or (nil? current-type)
+      (= current-type relaxed-type)
       (when-let [f (allowed-type-upgrades current-type)]
         (f relaxed-type))))
 
 (defn- changed-field->new-type
   "Given some fields and old and new types, filter out fields with unchanged types, then pair with the new types."
   [fields old-types new-types]
-  (let [new-if-changed #(when (not= %1 %2) %2)]
+  (let [new-if-changed #(when (and %1 (not= %1 %2)) %2)]
```

Comment:
that said maybe easier to parse if it's `(fn [old new] (when (and old ...)))`

Reply:
Yeah I think you're right that it's clearer with named args.

Given that you found the truth table helpful, maybe I'll hoist this to a private function and turn the table into a unit test.

Reply:
`new-if-changed` is coupled to line 625 because if the value is `nil` then it will be removed from the result.

I _think_ this would be slightly easier to understand if this were a reduction over `fields`, because then you could make this coupling more explicit.

Reply:
I gave it a try with a reduction, and avoiding a helper method, and this is what I got:

```clojure
(reduce 
   (fn [m [field old new]]
     (if (and old new (not= old new))
       (assoc m field new)
       m))
   {}
   (map vector fields old-types new-types))
```

Overall I find the existing code easier to read personally, and the "coupling" seems pretty thin to me. The nil values still have semantic meaning, its incidental that we want a map that filters them out.

Reply:
I agree on changing the whole approach. 

Since I'd already put some time into thinking about this implementation, and talked through it with Tim in our 1-1, I thought I'd still share the solution we'd both ended up liking the most.

```clojure
(defn- changed-field->new-type [fields old-types new-types]
  (let [updated?        (fn [o n] (and o (not= o n)))
        type-if-updated (fn [o n] (when (updated? o n) n))]
    (u/zipkeep fields (map type-if-updated old-types new-types))))
```

I suspect it's not to your taste, but interested in how and why.

The new `u/zipkeep` here is like `zipmap`, except it skips key-value pairs where either the key or value is nil. 

Not so sure about the name, also thought about `zipmap-keep` and `zipmap-some`. Just an optimization for `(m/filter-vals some? (zipmap ...))` in this instance. Naming is hard, so luckily we're not adding it today.